src/values/doc.ts

```diff
@@ -0,0 +1,85 @@
+import { JSONObject } from "../wire-protocol";
+import { TimeStub } from "./date-time";
+
+/**
+ * A Document Reference. Represents a pointer to a Document in the database. The
+ * Document may or may not exists.
+ */
+export class Ref {
```

Comment:
This doesn't differentiate between "named" references (e.g. Collections, Databases, Functions, Roles, AccessProviders, etc.) which are shaped like 

```json
{ "@ref": { "name": "Things", "coll": { "@mod": "Collection" } } }
```

Reply:
Can we apply a discriminated union type on top of `NamedDocument` to cover Collections vs Databases etc. or should we have subclasses for each?

Reply:
Also how do we handle user data on NamedDocument? Should that be generic?

Reply:
Started over from scratch. `Document` now always has the user-defined data at the top field, it just can't be used by typescript. Typescript users can cast to the provided `DocumentT` alias to extend `Document` with generic data.